Update pio and netcdf error checks

apcraig opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update pio and netcdf error checks
- [X] Developer(s): 
    anton-s, apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Full test suite run on Derecho cray, gnu, intel including io_suite, results look good, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#dc14ce98d71a742d9132d7ed8b7ecad3a39bd0a1
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please provide any additional information or relevant details below:

- Start with Anton's io_error branch on https://github.com/ACCESS-NRI/CICE.  Lots of work was done to add pio error checks.
- Update netcdf error checks leveraging nf90_strerror, added new subroutine ice_check_nc (similar to ice_pio_check in pio)
- Add USE_PIO1 cpp.  pio1 does not have pio_strerror so we need to compile that out.
- Fix a few minor bugs that were caught by the error checking, nothing that affects results.
- Add parallelio to conda_macos, not clear it works, but doesn't hurt netcdf builds/runs
- Update indentation in some files and update abort_ice calls to add file and line where needed.
- Cleaned up end of line whitespace via script
- Updated the github actions JRA55 gx3 forcing dataset to just download Jan 2005 data

Closes #858

apcraig commented:
I will update the test results when complete.

Recommend reviewing the changes with "ignore whitespace" turned on.  I ended up fixing a bunch of indentation issues (sorry about that, but it always feels good).  To do that, go to "Files Changed" and under settings click the box.

This represents the first phase of the IO upgrade, https://docs.google.com/document/d/1yNK1MV2v51MF_P_qz0Mva34ZxbGSwoPnNvXkm5Q_Hs0.  See also #862, #914

apcraig commented:
Full test suite run on Derecho cray, gnu, intel including io_suite, results look good, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#dc14ce98d71a742d9132d7ed8b7ecad3a39bd0a1

anton-seaice commented:
Thanks Craig - well done. Did you mean to update the docs? You said yes in the description - but there are no changes included. (Possibly USE_PIO1 should be included, although maybe its not 'officially' supported as it looks quite old). 

We could add a note about PIO2 vs PIO1 or supported/tested PIO versions in https://cice-consortium-cice.readthedocs.io/en/main/user_guide/ug_running.html#software-requirements ?

anton-seaice commented:
I think we can check for the specific error code here, and then use ice_check_nc if it is a different error code?:

https://github.com/CICE-Consortium/CICE/blob/dc14ce98d71a742d9132d7ed8b7ecad3a39bd0a1/cicecore/cicedyn/infrastructure/ice_grid.F90#L951

anton-seaice commented on cicecore/cicedyn/infrastructure/ice_read_write.F90:
```diff
@@ -1239,16 +1236,22 @@ subroutine ice_read_nc_xy(fid,  nrec,  varname, work,  diag, &
+         missingvalue = 1.234e30
+         status = nf90_get_att(fid, varid, "_FillValue", missingvalue)
+!          call ice_check_nc(status, subname//' ERROR: Missing _FillValue', &
+!                            file=__FILE__, line=__LINE__)
```
if (status != NC_ENOTATT) then
call ice_check_nc( ...

anton-seaice replied:
See the version in my suggestions :)

DeniseWorthen replied:
@apcraig  is the assignment of the missingvalue to 1.234e30 (which seems like an odd missingvalue) only to prevent uninitialized-variable errors under debug?

apcraig replied:
This is an odd section of code that I'm not sure what to do with.  This is only used with some diag flags, so maybe infrequently.  I think we want to avoid missingvalue being undefined but there isn't a "good" value to set it to by default.  You probably want it set to something that will make the mask ineffective in the diagnostics.  There are lots of other ways to do this including checking the return code on the _FillValue then not masking the diagnostics if missingvalue is not defined.  But I feel like this is a reasonable workaround for now.  We can always fix it if it ever becomes a problem.

DeniseWorthen replied:
That is fine w/ me. Thanks.

dabail10 replied:
I feel like we should be using spval or spval_dbl.

anton-seaice commented:
See suggestions here : https://github.com/anton-seaice/CICE/commit/d1eece938f2a21f89a58617091815e665a7a678f , it didnt let me merge with this PR

dabail10 commented:
I will try to take a look tomorrow. Today is a holiday for some of us in the USA (MLK day).

apcraig commented:
@anton-seaice, thanks for the review.  Lots of good comments.  I have updated the branch.  Will address your comments here.

- I cleaned up some long line lengths and some ' ERROR: ' syntax and error messages as suggested, just in the io/infrastructure routines though.   We try to keep things consistent throughout as much as we can, but it's always a work in progress (like indentation).
- I switched from colon to comma in the netcdf and pio error output.  That's a good idea.  I think it's OK to have an explicit message defined by CICE like "ERROR: ..." even if netcdf/pio will write "ERROR ...".  And I'm trying to add file and line numbers on the abort calls.  On an error, I think it's OK to lean towards extra and even redundant information in the error message especially if there is any chance it's easier to debug.
- I'm not inclined to use things like NC_ENOTATT unless they are absolutely necessary.  I think it adds a bit of unnecessary risk and complexity.  We only have a few places where we don't abort on an error and in those cases, I think it's fine to just check that there was a general error and take the next step, rather than differentiate for specific netcdf/pio errors.  Willing to reconsider if others feel differently.
- You suggested a couple changes, to remove a redundant set_pioerrorhandling and to remove some output when opening a file, both in ice_pio.F90.  I think both of those are OK as is.  The pioerrorhandling redundancy is for CESMCOUPLED and I'm happy to let them manage this setting.  The writing of the file opening is probably somewhat useful overall.
- Regarding adding parallelio to the conda_macos env.  Did you get this to work?  Everything works fine for me when using netcdf (with the conda parallelio + -lpiof, but if I setup a case with pio2 on my mac (at home), it dumps core very early.  I think it's OK to include parallelio for conda_macos as it doesn't seem to cause problems with netcdf, but if it doesn't work with pio, maybe we shouldn't include it?  Along those lines, I modified the Macros file for conda_macos to turn on the -lpiof -lpioc only when pio2 is chosen just to make it even less likely that the parallelio will cause a problem later.  But am curious what your experience is with parallelio with conda_macos.

anton-seaice commented:
A few more ERROR messages to tidyup: 

https://github.com/anton-seaice/CICE/commit/40a56b375aef0f6fcc7bf0771db8b9e4329a4493

apcraig commented:
Thanks @anton-seaice, I cherry-picked your latest modifications.  There are always more things to fix like this.  Also, I meant to mention that while you can't push to my branches, you can PR to them.  Generally, that's what we do if we're collaborating on development.  Someone sets up a branch and others PR to it.  Feel free to do that if it's useful at times.

I have not worked to improve

```
call pio_seterrorhandling(ice_pio_subsystem, PIO_RETURN_ERROR)
call pio_seterrorhandling(ice_pio_subsystem, PIO_INTERNAL_ERROR)
```

I think what we want is for the default to be "internal error" and for us to switch to "return error" in the sections where we actually check errors.  In practice, we could probably just set "internal error" everywhere and let pio abort for us not worry about return codes too much.  Then we would use "return error" ONLY when we want to ignore an error.  Anyway, there are various appropriate strategies.  But now that we've added all the error checking in our code, we have more options.  The one important issue is when pio doesn't return error codes.  I think we set "internal error" around those, but we certainly could look at all that more carefully.

apcraig commented:
Also, glad to hear the pio is working for you with conda macos.  I'm using an older version of the OS (and maybe hardware), so that might be part of my problem.  If it works for you, I'm happy to leave it in, especially as it doesn't seem to impact the netcdf option even if the pio option doesn't work all the time.  It's kind of cool there is conda pio support now, thanks for adding it.

DeniseWorthen commented:
@apcraig We need to make a change in the cmeps driver related to #921 in order to compile and test this.

```
diff --git a/cicecore/drivers/nuopc/cmeps/ice_mesh_mod.F90 b/cicecore/drivers/nuopc/cmeps/ice_mesh_mod.F90
index 9493add..0c430f6 100644
--- a/cicecore/drivers/nuopc/cmeps/ice_mesh_mod.F90
+++ b/cicecore/drivers/nuopc/cmeps/ice_mesh_mod.F90
@@ -437,7 +437,8 @@ subroutine ice_mesh_init_tlon_tlat_area_hm()

     use ice_grid      , only : tlon, tlat, hm, tarea, ULON, ULAT, HTN, HTE, ANGLE, ANGLET
     use ice_grid      , only : uarea, uarear, tarear!, tinyarea
-    use ice_grid      , only : dxT, dyT, dxU, dyU, dyhx, dxhy, cyp, cxp, cym, cxm
+    use ice_grid      , only : dxT, dyT, dxU, dyU
+    use ice_dyn_shared, only : dxhy, dyhx, cxp, cyp, cxm, cym
     use ice_grid      , only : makemask
     use ice_boundary  , only : ice_HaloUpdate
     use ice_domain    , only : blocks_ice, nblocks, halo_info, distrb_info
```

DeniseWorthen commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_history_write.F90:
```diff
@@ -137,654 +138,653 @@ subroutine ice_write_hist (ns)
+                              file=__FILE__, line=__LINE__)
+         elseif (use_leap_years) then
+            status = nf90_put_att(ncid,varid,'calendar','Gregorian')
+            call ice_check_nc(status, subname// ' ERROR: time calendar', &
```
```suggestion
            call ice_check_nc(status, subname// ' ERROR: time calendar Gregorian', &
```

DeniseWorthen commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_history_write.F90:
```diff
@@ -137,654 +138,653 @@ subroutine ice_write_hist (ns)
+                              file=__FILE__, line=__LINE__)
+         elseif (days_per_year == 365 .and. .not.use_leap_years ) then
+            status = nf90_put_att(ncid,varid,'calendar','noleap')
+            call ice_check_nc(status, subname// ' ERROR: time calendar', &
```
```suggestion
            call ice_check_nc(status, subname// ' ERROR: time calendar no leap', &
```

DeniseWorthen commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_history_write.F90:
```diff
@@ -137,654 +138,653 @@ subroutine ice_write_hist (ns)
+
+         if (days_per_year == 360) then
+            status = nf90_put_att(ncid,varid,'calendar','360_day')
+            call ice_check_nc(status, subname// ' ERROR: time calendar', &
```
```suggestion
            call ice_check_nc(status, subname// ' ERROR: time calendar 360_day', &
```

DeniseWorthen commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_history_write.F90:
```diff
@@ -137,654 +138,653 @@ subroutine ice_write_hist (ns)
+                              file=__FILE__, line=__LINE__)
+            if (days_per_year == 360) then
+               status = nf90_put_att(ncid,varid,'calendar','360_day')
+               call ice_check_nc(status, subname// ' ERROR: time calendar1', &
```
As above, here you're calling them calendar1, 2 and 3. So perhaps make them all align w/ the calendar you're retrieving?

DeniseWorthen commented on cicecore/cicedyn/infrastructure/io/io_pio2/ice_history_write.F90:
```diff
@@ -192,73 +191,106 @@ subroutine ice_write_hist (ns)
+
+         if (days_per_year == 360) then
+            call ice_pio_check(pio_put_att(File,varid,'calendar','360_day'), &
+                 subname//' ERROR: defining att calendar1 tb',file=__FILE__,line=__LINE__)
```
I'm not sure what "tb" is supposed to mean, but I would suggest the errors call out which calendar is being checked (360_day, noleap and Gregorian).

apcraig commented:
Thanks @DeniseWorthen.  I have updated the calendar error messages.  I also update the nuopc driver mesh file.  The mesh file changes are related to a prior code mod, but happy to include it here.  We try to keep the driver code up to date with changes to the standalone model, but don't always succeed.  Let us know anytime if we can help resolve anything like that when it comes up.

anton-seaice commented:
> I think what we want is for the default to be "internal error" and for us to switch to "return error" in the sections where we actually check errors. 

Ok - this is what I have done in this change.

anton-seaice commented:
> Thanks @DeniseWorthen. I have updated the calendar error messages. I also update the nuopc driver mesh file. The mesh file changes are related to a prior code mod, but happy to include it here. We try to keep the driver code up to date with changes to the standalone model, but don't always succeed. Let us know anytime if we can help resolve anything like that when it comes up.

I compiled, including these changes, using the nuopc driver and CESMCOUPLED set, and found one typo in ice_pio - sent you a PR Tony.

dabail10 commented on cicecore/cicedyn/infrastructure/ice_grid.F90:
```diff
@@ -224,7 +224,7 @@ subroutine alloc_grid
          hm       (nx_block,ny_block,max_blocks), & ! land/boundary mask, thickness (T-cell)
          bm       (nx_block,ny_block,max_blocks), & ! task/block id
-         uvm      (nx_block,ny_block,max_blocks), & ! land/boundary mask, velocity (U-cell) - water in case of all water point
+         uvm      (nx_block,ny_block,max_blocks), & ! land/boundary mask, velocity (U-cell) - all water point
```
Sorry, but I don't understand this comment. Are we trying to say that uvm is only ice-covered points with aice > puny?

apcraig replied:
I don't understand it either and didn't try to, just wanted to shorten the line.  I think it's a u mask where all u points are surrounded by non-land t points.  If someone wants to suggest an updated comment, I'm happy to change it.

eclare108213 replied:
That comment came in with some of the C-grid stuff.  Please remove the 'water' part.  I think what it means is that uvm = 1 only when all of the surrounding cells are not land (which is the definition of uvm all the time).

dabail10 commented on cicecore/cicedyn/infrastructure/ice_read_write.F90:
```diff
@@ -1239,16 +1236,22 @@ subroutine ice_read_nc_xy(fid,  nrec,  varname, work,  diag, &
 !            write(nu_diag,*) subname,' Dim name = ',trim(dimname),', size = ',dimlen
 !         enddo
+         ! optional
+         missingvalue = 1.234e30
```
Should we be hard-coding missingvalue here?

apcraig replied:
@DeniseWorthen asked a similar question.  My response was

This is an odd section of code that I'm not sure what to do with. This is only used with some diag flags, so maybe infrequently. I think we want to avoid missingvalue being undefined but there isn't a "good" value to set it to by default. You probably want it set to something that will make the mask ineffective in the diagnostics. There are lots of other ways to do this including checking the return code on the _FillValue then not masking the diagnostics if missingvalue is not defined. But I feel like this is a reasonable workaround for now. We can always fix it if it ever becomes a problem.

dabail10 commented on cicecore/cicedyn/infrastructure/io/io_pio2/ice_history_write.F90:
```diff
@@ -192,73 +191,106 @@ subroutine ice_write_hist (ns)
+      endif
+
+      call ice_pio_check(pio_def_dim(File,'ni',nx_global,imtid), &
+           subname//' ERROR: defining dim ni',file=__FILE__,line=__LINE__)
```
I guess this is ok. PIO has it's own error checking routines internally.

apcraig replied:
Regarding your comment below, the pio error checking turns off the internal error handling for the most part.  There are some pio calls that don't return an error code, for those pio internal error is on.

dabail10 reviewed (commented):
This looks good overall. I think @DeniseWorthen and @apcraig have already provided lots of feedback. I think in PIO if we want to do the ice_pio_check, we should turn off the internal PIO error handling.

apcraig commented:
> This looks good overall. I think @DeniseWorthen and @apcraig have already provided lots of feedback. I think in PIO if we want to do the ice_pio_check, we should turn off the internal PIO error handling.

apcraig commented:
Any additional review comments.  If not, I'll run a final test suite and then maybe we can merge before the weekend?  Thanks.

phil-blain commented:
Hi Tony, I haven't had the time yet to review this branch. I'd like to take a look. I'll try to do it this afternoon.

apcraig commented:
Thanks @phil-blain, I'm happy to delay the merge to next week too.  Just looking to start the second phase of the IO updates.  No rush though.

phil-blain reviewed (commented):
Thanks for this work Tony, I'm about halfway done through the changes but I'll post what I noted so far now, since I'm about to leave. I'll finish my review tomorrow morning.

Regarding the new subroutine `ice_check_nc`, I would have suggested using a function instead, since this way we could wrap the NetCDF calls, which adds less lines of code, i.e. instead of:
~~~fortran
        status = nf90_put_att(ncid,nf90_global,'io_flavor','io_netcdf')
        call ice_check_nc(status, subname// ' ERROR: global attribute io_flavor', &
                          file=__FILE__, line=__LINE__)
~~~

we could do:

~~~fortran
status = ice_check_nc(nf90_put_att(ncid,nf90_global,'io_flavor','io_netcdf'), subname// ' ERROR: global attribute io_flavor', &
                      file=__FILE__, line=__LINE__))
~~~

but I do not have a strong preference (anyway the code is already update to use the first approach!)

**EDIT** Re-reading my self, I realize we can also use the same approach by wrapping the NetCDF calls in the call to `ice_check_nc`, this is mostly what is done for the PIO code. I must have been thinking of something else... but anyway as I wrote the code is already updated and I think it's OK this way.

phil-blain commented on doc/source/user_guide/ug_running.rst:
```diff
@@ -16,59 +16,88 @@ To run stand-alone, CICE requires
-- NetCDF (this is actually optional but required to test out of the box configurations)
-- MPI (this is actually	optional but without it	you can	only run on 1 processor)
+- Fortran and C	compilers (Intel, PGI, GNU, Cray, NVHPC, AOCC, and NAG have been tested)
+- NetCDF (optional, but required to test standard configurations that have netcdf grid, input, and forcing files)
```
```suggestion
- NetCDF (optional, but required to test standard configurations that have netCDF grid, input, and forcing files)
```

phil-blain commented on doc/source/user_guide/ug_running.rst:
```diff
@@ -16,59 +16,88 @@ To run stand-alone, CICE requires
 point, the Consortium is regularly testing on several different compilers, but not 
-necessarily on all possible versions or combinations.  A CICE goal is to be relatively portable
+necessarily on all possible versions or combinations.  CICE supports both PIO1 and PIO2.  To
+use PIO1, the cpp USE_PIO1 should also be set.  A CICE goal is to be relatively portable
```
```suggestion
use PIO1, the ``USE_PIO1`` macro should also be set.  A CICE goal is to be relatively portable
```

phil-blain commented on doc/source/user_guide/ug_running.rst:
```diff
@@ -16,59 +16,88 @@ To run stand-alone, CICE requires
 - NAG 6.2
+- NVC 23.5-0
 
 The Consortium has tested the following mpi versions,
```
```suggestion
The Consortium has tested the following MPI implementations and versions,
```

while we're at it!

phil-blain commented on cicecore/cicedyn/infrastructure/io/io_pio2/ice_history_write.F90:
```diff
@@ -1308,26 +1425,30 @@ end subroutine ice_write_hist_attrs
-      use ice_kinds_mod
-      use ice_pio
-      use pio
+      ! use ice_kinds_mod, only: int_kind
```
if we do not need this why not delete it instead of commenting it?

phil-blain reviewed (commented):
Thanks Tony and Anton for this work. this was surely a lot of grunt work but I think it's really valuable to output the NetCDF error string  when NetCDF calls fail, it helped me understand what was going more quickly on a few times in the past. 

I left a few suggestions below. 

One thing I though about is how to enforce the use of `ice_check_nc` and `ice_pio_check` for new code, apart from careful reviewing... I can't think of an easy and robust way to automate this...

phil-blain commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_restart.F90:
```diff
@@ -830,13 +846,15 @@ subroutine final_restart()
-      if (my_task == master_task) &
-         write(nu_diag,'(a,i8,4x,i4.4,a,i2.2,a,i2.2,a,i5.5)') 'Restart read/written ',istep1,myear,'-',mmonth,'-',mday,'-',msec
-
+      if (my_task == master_task) then
+         ! ncid is only valid on master
+         status = nf90_close(ncid)
+         call ice_check_nc(status, subname//' ERROR: closing', file=__FILE__, line=__LINE__)
+         write(nu_diag,'(a,i8,4x,i4.4,a,i2.2,a,i2.2,a,i5.5)') &
+            'Restart read/written ',istep1,myear,'-',mmonth,'-',mday,'-',msec
+      endif
```
Also, these changes fix https://github.com/CICE-Consortium/CICE/issues/858 so we could add 
> `Closes: https://github.com/CICE-Consortium/CICE/issues/858`

to the PR description or the commit message for this change :) so the issue gets closed when we merge.

apcraig commented:
@phil-blain.  Thanks for your review.  I have made the updates you suggest.  Let me also add some discussion.

The question of whether to use

```
 status = nf90_put_att(ncid,nf90_global,'io_flavor','io_netcdf')
 call ice_check_nc(status, subname// ' ERROR: global attribute io_flavor', file=__FILE__, line=__LINE__)
```

or

```
call ice_check_nc(nf90_put_att(ncid,nf90_global,'io_flavor','io_netcdf'), subname// ' ERROR: global attribute io_flavor', file=__FILE__, line=__LINE__)
```

is a good one.  I thought about it before updating the implementation.  The pio code uses the latter and the netcdf code uses the former.  For me, I prefer the former as I find it much more readable, especially if the netcdf has to be debugged.  I agree that either way works.  **Maybe a bigger question is whether the pio and netcdf codes should use the same strategy.**  At this point, I think we can leave as is unless someone feels strongly we should use the same error checking approach in pio and netcdf.

With respect to maintaining the error checks, this is a bit of a challenge.   Checking the icepack abort code consistently is another similar "requirement" as is use of "subname" and so forth.  We just have to try to remember to review these things.  Also, periodically, if I see an problem, I go through the code quickly and update as needed.  Not ideal, but I don't know if we can automate.

I added a close #858 to the PR (in the PR description above).

I think I've addressed everyone's comments up to this point, feel free to review again and/or let me know if I've missed something.

apcraig commented on cicecore/cicedyn/infrastructure/ice_domain.F90:
```diff
@@ -487,14 +488,27 @@ subroutine init_domain_distribution(KMTG,ULATG,grid_ice)
 #ifdef USE_NETCDF
          status = nf90_open(distribution_wght_file, NF90_NOWRITE, fid)
          if (status /= nf90_noerr) then
-            call abort_ice (subname//'ERROR: Cannot open '//trim(distribution_wght_file))
+            call abort_ice(subname//' ERROR: Cannot open '//trim(distribution_wght_file), &
+                           file=__FILE__, line=__LINE__)
```
Correct.

apcraig commented on cicecore/cicedyn/infrastructure/io/io_netcdf/ice_restart.F90:
```diff
@@ -763,54 +774,59 @@ subroutine write_restart_field(nu,nrec,work,atype,vname,ndim3,diag)
       character(len=*), parameter :: subname = '(write_restart_field)'
 
 #ifdef USE_NETCDF
+      varid = -99
```
Correct.

apcraig commented:
Also, I changed the missingvalue=1.234e30 to spval_dbl.  I was thinking we wanted to avoid spval_dbl as the default because it might actually mean something in the fields that we might want to avoid (like maybe land vs water), but decided, after many questions, that it makes sense to use spval_dbl there.

phil-blain commented on doc/source/developer_guide/dg_about.rst:
```diff
@@ -73,5 +73,11 @@ Overall, CICE code should be implemented as follows,
        call icepack_warnings_flush(nu_diag)
        if (icepack_warnings_aborted()) call abort_ice(error_message=subname, file=__FILE__, line=__LINE__)
 
+  * Use ``ice_check_nc`` or ``ice_pio_check`` after netcdf or pio calls to check for return errors.
+
+  * Use subroutine ``abort_ice`` to abort the model run. Do not use stop or MPI_ABORT.  Use optional arguments (file=__FILE__, line=__LINE__) in calls to ``abort_ice`` to improve debugging
+
+  * Write output to stdout from the master task only unless the output is associated with an abort call.  Write to unit ``nu_diag`` following the current standard.  Do not use units 5 or 6.  Do not use the print statement.
+
```
nice additions!

anton-seaice commented on doc/source/developer_guide/dg_about.rst:
```diff
@@ -73,5 +73,11 @@ Overall, CICE code should be implemented as follows,
        call icepack_warnings_flush(nu_diag)
        if (icepack_warnings_aborted()) call abort_ice(error_message=subname, file=__FILE__, line=__LINE__)
 
+  * Use ``ice_check_nc`` or ``ice_pio_check`` after netcdf or pio calls to check for return errors.
```
Should we make a note about setting/unsetting PIO_RETURN_ERROR around ice_pio_check too ? ice_pio_check won't do anything otherwise.

apcraig replied:
Personally, I think this is a little too much detail to be in the documentation.  It's a tough balance between making the documentation useful versus overwhelming.  From my perspective, I hope folks will read the coding suggestions in the documentation (still not clear how often that happens), but also mostly follow current coding approaches.  If we want to add more details in the documentation, I'm open to that.  But we also need to maintain it.

anton-seaice commented:
> With respect to maintaining the error checks, this is a bit of a challenge. Checking the icepack abort code consistently is another similar "requirement" as is use of "subname" and so forth. We just have to try to remember to review these things. Also, periodically, if I see an problem, I go through the code quickly and update as needed. Not ideal, but I don't know if we can automate.

We could add a check item to the PR checklist? (i.e. do the changes follow with coding standard outlined in the documentation, including defining subname for each subrountine and function and calling icepack_warnings_flush after icepack calls?)

Also - there might be a bot that could respond to each PR with a reviever checklist (which would have the items from the coding standard on it).

DeniseWorthen reviewed (approved):
Good work. Thanks.

eclare108213 reviewed (approved):
wow
